File: benchmarks-QuantumOptics.jl/variance_state.jl
using QuantumOptics
using BenchmarkTools
include("benchmarkutils.jl")

name = "variance_state"

samples = 5
evals = 100
cutoffs = [5000:5000:100000;]

function setup(N)
    b = FockBasis(N-1)
    op = (destroy(b) + create(b))
    psi = Ket(b, ones(Complex128, N)/sqrt(N))
    op, psi
end

function f(op, psi)
    variance(op, psi)
end

println("Benchmarking: ", name)
print("Cutoff: ")
checks = Dict{Int, Float64}()
results = []
for N in cutoffs
    print(N, " ")
    op, psi = setup(N)
    checks[N] = abs(f(op, psi))
    t = @belapsed f($op, $psi) samples=samples evals=evals
    push!(results, Dict("N"=>N, "t"=>t))
end
println()

benchmarkutils.check(name, checks)
benchmarkutils.save(name, results)

File: benchmarks-QuTiP/variance_state.py
import qutip as qt
import numpy as np
import benchmarkutils

name = "variance_state"

samples = 5
evals = 100
cutoffs = range(5000, 100001, 5000)

def f(op, state):
    return qt.variance(op, state)

print("Benchmarking:", name)
print("Cutoff: ", end="", flush=True)
checks = {}
results = []
for N in cutoffs:
    print(N, "", end="", flush=True)
    op = (qt.destroy(N) + qt.create(N))
    psi = qt.Qobj(np.ones(N, complex)/(N**(1/2)))
    checks[N] = f(op, psi)
    t = benchmarkutils.run_benchmark(f, op, psi, samples=samples, evals=evals)
    results.append({"N": N, "t": t})
print()

benchmarkutils.check(name, checks)
benchmarkutils.save(name, results)
